File: gpui/crates/azem-ipc/src/codec.rs
use std::io;

use bytes::Bytes;
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};

use crate::generated::{BinaryMetadata, Envelope, MAX_BINARY_CHUNK_BYTES, MAX_CONTROL_FRAME_BYTES};

const WIRE_CONTROL: u8 = 1;
const WIRE_BINARY: u8 = 2;
const MAX_BINARY_HEADER_BYTES: usize = 64 << 10;

#[derive(Debug)]
pub enum Frame {
    Control(Box<Envelope>),
    Binary(BinaryMetadata, Bytes),
}

pub async fn read_frame<R>(reader: &mut R) -> io::Result<Frame>
where
    R: AsyncRead + Unpin + ?Sized,
{
    let size = reader.read_u32().await? as usize;
    if size == 0 || size > MAX_CONTROL_FRAME_BYTES {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("invalid IPC frame size {size}"),
        ));
    }
    let mut payload = vec![0_u8; size];
    reader.read_exact(&mut payload).await?;
    match payload[0] {
        WIRE_CONTROL => {
            let envelope = serde_json::from_slice(&payload[1..])
                .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))?;
            Ok(Frame::Control(Box::new(envelope)))
        }
        WIRE_BINARY => {
            if payload.len() < 3 {
                return Err(io::Error::new(
                    io::ErrorKind::UnexpectedEof,
                    "truncated IPC binary frame",
                ));
            }
            let header_len = u16::from_be_bytes([payload[1], payload[2]]) as usize;
            if header_len == 0
                || header_len > MAX_BINARY_HEADER_BYTES
                || 3 + header_len > payload.len()
            {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    "invalid IPC binary header",
                ));
            }
            let metadata = serde_json::from_slice(&payload[3..3 + header_len])
                .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))?;
            let data = Bytes::copy_from_slice(&payload[3 + header_len..]);
            if data.len() > MAX_BINARY_CHUNK_BYTES {
                return Err(io::Error::new(
                    io::ErrorKind::InvalidData,
                    "IPC binary chunk is too large",
                ));
            }
            Ok(Frame::Binary(metadata, data))
        }
        kind => Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("unsupported IPC wire kind {kind}"),
        )),
    }
}

pub async fn write_envelope<W>(writer: &mut W, envelope: &Envelope) -> io::Result<()>
where
    W: AsyncWrite + Unpin + ?Sized,
{
    let encoded = serde_json::to_vec(envelope)
        .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))?;
    write_physical(writer, WIRE_CONTROL, &[], &encoded).await
}

pub async fn write_binary<W>(
    writer: &mut W,
    metadata: &BinaryMetadata,
    data: &[u8],
) -> io::Result<()>
where
    W: AsyncWrite + Unpin + ?Sized,
{
    if data.len() > MAX_BINARY_CHUNK_BYTES {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "IPC binary chunk is too large",
        ));
    }
    let header = serde_json::to_vec(metadata)
        .map_err(|error| io::Error::new(io::ErrorKind::InvalidData, error))?;
    if header.len() > MAX_BINARY_HEADER_BYTES {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "IPC binary metadata is too large",
        ));
    }
    let prefix = (header.len() as u16).to_be_bytes();
    let mut complete_header = Vec::with_capacity(prefix.len() + header.len());
    complete_header.extend_from_slice(&prefix);
    complete_header.extend_from_slice(&header);
    write_physical(writer, WIRE_BINARY, &complete_header, data).await
}

async fn write_physical<W>(
    writer: &mut W,
    kind: u8,
    header: &[u8],
    payload: &[u8],
) -> io::Result<()>
where
    W: AsyncWrite + Unpin + ?Sized,
{
    let size = 1 + header.len() + payload.len();
    if size > MAX_CONTROL_FRAME_BYTES {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "IPC physical frame is too large",
        ));
    }
    writer.write_u32(size as u32).await?;
    writer.write_u8(kind).await?;
    writer.write_all(header).await?;
    writer.write_all(payload).await?;
    writer.flush().await
}

#[cfg(test)]
mod tests {
    use super::{Frame, read_frame, write_binary, write_envelope};
    use crate::{BinaryMetadata, Method, generated::MAX_CONTROL_FRAME_BYTES, protocol::envelope};

    #[tokio::test]
    async fn control_and_binary_frames_round_trip() {
        let (mut writer, mut reader) = tokio::io::duplex(4096);
        let task = tokio::spawn(async move {
            let mut request = envelope("request");
            request.id = "request-1".into();
            request.method = Some(Method::Initialise);
            write_envelope(&mut writer, &request).await.unwrap();
            let metadata = BinaryMetadata {
                transfer_id: "transfer-1".into(),
                purpose: "attachment".into(),
                count: 1,
                byte_length: 3,
                ..Default::default()
            };
            write_binary(&mut writer, &metadata, b"abc").await.unwrap();
        });

        match read_frame(&mut reader).await.unwrap() {
            Frame::Control(envelope) => {
                assert_eq!(envelope.id, "request-1");
                assert_eq!(envelope.method, Some(Method::Initialise));
            }
            frame => panic!("unexpected frame: {frame:?}"),
        }
        match read_frame(&mut reader).await.unwrap() {
            Frame::Binary(metadata, data) => {
                assert_eq!(metadata.transfer_id, "transfer-1");
                assert_eq!(&data[..], b"abc");
            }
            frame => panic!("unexpected frame: {frame:?}"),
        }
        task.await.unwrap();
    }

    #[tokio::test]
    async fn oversized_frame_is_rejected_before_payload_read() {
        let (mut writer, mut reader) = tokio::io::duplex(16);
        tokio::spawn(async move {
            use tokio::io::AsyncWriteExt;
            writer
                .write_all(&((MAX_CONTROL_FRAME_BYTES as u32) + 1).to_be_bytes())
                .await
                .unwrap();
        });
        let error = read_frame(&mut reader).await.unwrap_err();
        assert_eq!(error.kind(), std::io::ErrorKind::InvalidData);
    }
}

File: internal/desktopipc/codec.go
package desktopipc

import (
	"bufio"
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

const (
	wireControl          byte = 1
	wireBinary           byte = 2
	maxBinaryHeaderBytes      = 64 << 10
)

type DecodedFrame struct {
	Envelope *Envelope
	Binary   *BinaryMetadata
	Data     []byte
}

type Codec struct {
	reader *bufio.Reader
	writer *bufio.Writer
	mu     sync.Mutex
}

func NewCodec(stream io.ReadWriter) *Codec {
	return &Codec{reader: bufio.NewReaderSize(stream, 64<<10), writer: bufio.NewWriterSize(stream, 64<<10)}
}

func (codec *Codec) ReadFrame() (DecodedFrame, error) {
	var sizeBuffer [4]byte
	if _, err := io.ReadFull(codec.reader, sizeBuffer[:]); err != nil {
		return DecodedFrame{}, err
	}
	size := int(binary.BigEndian.Uint32(sizeBuffer[:]))
	if size < 1 || size > MaxControlFrameBytes {
		return DecodedFrame{}, fmt.Errorf("IPC physical frame size %d is invalid", size)
	}
	payload := make([]byte, size)
	if _, err := io.ReadFull(codec.reader, payload); err != nil {
		return DecodedFrame{}, err
	}
	switch payload[0] {
	case wireControl:
		var envelope Envelope
		if err := json.Unmarshal(payload[1:], &envelope); err != nil {
			return DecodedFrame{}, fmt.Errorf("decode IPC control frame: %w", err)
		}
		if err := envelope.Validate(); err != nil {
			return DecodedFrame{}, err
		}
		return DecodedFrame{Envelope: &envelope}, nil
	case wireBinary:
		if len(payload) < 3 {
			return DecodedFrame{}, errors.New("IPC binary frame is truncated")
		}
		headerLength := int(binary.BigEndian.Uint16(payload[1:3]))
		if headerLength <= 0 || headerLength > maxBinaryHeaderBytes || 3+headerLength > len(payload) {
			return DecodedFrame{}, errors.New("IPC binary frame header is invalid")
		}
		var metadata BinaryMetadata
		if err := json.Unmarshal(payload[3:3+headerLength], &metadata); err != nil {
			return DecodedFrame{}, fmt.Errorf("decode IPC binary metadata: %w", err)
		}
		data := payload[3+headerLength:]
		if len(data) > MaxBinaryChunkBytes {
			return DecodedFrame{}, fmt.Errorf("IPC binary chunk exceeds %d bytes", MaxBinaryChunkBytes)
		}
		if metadata.TransferID == "" || metadata.Index < 0 || metadata.Count < 0 {
			return DecodedFrame{}, errors.New("IPC binary metadata is incomplete")
		}
		return DecodedFrame{Binary: &metadata, Data: data}, nil
	default:
		return DecodedFrame{}, fmt.Errorf("unsupported IPC wire frame type %d", payload[0])
	}
}

func (codec *Codec) WriteEnvelope(envelope Envelope) error {
	if err := envelope.Validate(); err != nil {
		return err
	}
	encoded, err := json.Marshal(envelope)
	if err != nil {
		return err
	}
	return codec.writeFrame(wireControl, nil, encoded)
}

func (codec *Codec) WriteBinary(metadata BinaryMetadata, data []byte) error {
	if metadata.TransferID == "" {
		return errors.New("IPC binary transfer id is required")
	}
	if len(data) > MaxBinaryChunkBytes {
		return fmt.Errorf("IPC binary chunk exceeds %d bytes", MaxBinaryChunkBytes)
	}
	header, err := json.Marshal(metadata)
	if err != nil {
		return err
	}
	if len(header) > maxBinaryHeaderBytes {
		return errors.New("IPC binary metadata is too large")
	}
	var prefix [2]byte
	binary.BigEndian.PutUint16(prefix[:], uint16(len(header)))
	return codec.writeFrame(wireBinary, append(prefix[:], header...), data)
}

func (codec *Codec) WriteTerminalReplay(response Envelope, chunks []terminalChunk) error {
	if err := response.Validate(); err != nil {
		return err
	}
	encoded, err := json.Marshal(response)
	if err != nil {
		return err
	}
	headers := make([][]byte, len(chunks))
	for index, chunk := range chunks {
		if chunk.metadata.TransferID == "" || len(chunk.data) > MaxBinaryChunkBytes {
			return errors.New("IPC terminal replay chunk is invalid")
		}
		header, headerErr := json.Marshal(chunk.metadata)
		if headerErr != nil {
			return headerErr
		}
		if len(header) > maxBinaryHeaderBytes {
			return errors.New("IPC binary metadata is too large")
		}
		var prefix [2]byte
		binary.BigEndian.PutUint16(prefix[:], uint16(len(header)))
		headers[index] = append(prefix[:], header...)
	}
	codec.mu.Lock()
	defer codec.mu.Unlock()
	if err := codec.writeFrameLocked(wireControl, nil, encoded); err != nil {
		return err
	}
	for index, chunk := range chunks {
		if err := codec.writeFrameLocked(wireBinary, headers[index], chunk.data); err != nil {
			return err
		}
	}
	return nil
}

func (codec *Codec) writeFrame(kind byte, header, payload []byte) error {
	codec.mu.Lock()
	defer codec.mu.Unlock()
	return codec.writeFrameLocked(kind, header, payload)
}

func (codec *Codec) writeFrameLocked(kind byte, header, payload []byte) error {
	size := 1 + len(header) + len(payload)
	if size > MaxControlFrameBytes {
		return fmt.Errorf("IPC physical frame exceeds %d bytes", MaxControlFrameBytes)
	}
	var sizeBuffer [4]byte
	binary.BigEndian.PutUint32(sizeBuffer[:], uint32(size))
	if _, err := codec.writer.Write(sizeBuffer[:]); err != nil {
		return err
	}
	if err := codec.writer.WriteByte(kind); err != nil {
		return err
	}
	if len(header) > 0 {
		if _, err := codec.writer.Write(header); err != nil {
			return err
		}
	}
	if len(payload) > 0 {
		if _, err := codec.writer.Write(payload); err != nil {
			return err
		}
	}
	return codec.writer.Flush()
}
